visualization/app/codeCharta/util/fileValidator.ts

```diff
@@ -134,3 +151,70 @@ function getAsApiVersion(version: string): ApiVersion {
 		minor: Number(version.split(".")[1])
 	}
 }
+
+function validateFixedFolders(file: ExportCCFile, result: CCValidationResult) {
+	const notFixed: string[] = []
+	const outOfBounds: string[] = []
+	const intersections: Set<string> = new Set()
+
+	file.nodes[0].children.forEach(node => {
+		if (node.fixed === undefined) {
+			notFixed.push(node.name)
+		} else {
+			if (isOutOfBounds(node)) {
+				outOfBounds.push(node.name)
+			}
+
+			file.nodes[0].children.forEach(node2 => {
+				if (node !== node2 && rectanglesIntersect(node.fixed, node2.fixed)) {
+					if (!intersections.has(`${node2.name} and ${node.name}`)) {
+						intersections.add(`${node.name} and ${node2.name}`)
+					}
+				}
+			})
+		}
+	})
+
+	if (notFixed.length > 0 && notFixed.length !== file.nodes[0].children.length) {
+		result.title = ERROR_MESSAGES.notAllFoldersAreFixed.title
+		result.error.push(`${ERROR_MESSAGES.notAllFoldersAreFixed.message} Found: ${notFixed.join(", ")}`)
+	}
+
+	if (outOfBounds.length > 0) {
+		result.title = ERROR_MESSAGES.fixedFoldersOutOfBounds.title
+		result.error.push(`${ERROR_MESSAGES.fixedFoldersOutOfBounds.message} Found: ${outOfBounds.join(", ")}`)
+	}
+
+	if (intersections.size > 0) {
+		result.title = ERROR_MESSAGES.fixedFoldersOverlapped.title
+		result.error.push(`${ERROR_MESSAGES.fixedFoldersOverlapped.message} Found: ${[...intersections.values()].join(", ")}`)
+	}
+}
+
+function rectanglesIntersect(rect1: Fixed, rect2: Fixed): boolean {
+	if (rect1 !== undefined && rect2 !== undefined) {
+		return (
+			isInRectangle(rect1.x, rect1.y, rect2) ||
+			isInRectangle(rect1.x, rect1.y + rect1.height, rect2) ||
+			isInRectangle(rect1.x + rect1.width, rect1.y, rect2) ||
+			isInRectangle(rect1.x + rect1.width, rect1.y + rect1.height, rect2)
+		)
+	}
+}
+
+function isInRectangle(x: number, y: number, rect: Fixed): boolean {
+	return x >= rect.x && x <= rect.x + rect.width && y >= rect.y && y <= rect.y + rect.height
+}
```

Comment:
What about using `isOverlapping`? It pretty much does the same as your code but it does seem a bit nicer to read and less checks:

```js
function isOverlapping(first, second) {
  return (
    first.y >= second.y + second.height &&
    first.y + first.height <= second.y &&
    first.x <= second.x + second.width &&
    first.x + first.width >= second.x
  )
}
```

(The logic should be correct but please check again)

Reply:
I tried multiple of these overlapping functions, but none seem to work. This is not working either for me.

Reply:
Let's get the first branch merged and everything else ready to have a look at this at the end again.

Reply:
I tried this [solution](https://www.geeksforgeeks.org/find-two-rectangles-overlap/), but it's just not working.

```ts
export function isOverlapping(rect1: FixedPosition, rect2: FixedPosition): boolean {
    // l = left top corner
    // r = right bottom corner
    const l1 = {x: rect1.left , y: rect1.top}
    const r1 = {x: rect1.left + rect1.width, y: rect1.top + rect1.height}

    const l2 = {x: rect2.left , y: rect2.top}
    const r2 = {x: rect2.left + rect2.width, y: rect2.top + rect2.height}

    {
        // If one rectangle is on left side of other
        if (l1.x >= r2.x || l2.x >= r1.x)
            return false;

        // If one rectangle is above other
        if (l1.y <= r2.y || l2.y <= r1.y)
            return false;

        return true;
    }
}
```

```ts
it("should show that rect1 and rect2 are overlapping", () => {
            const rect1: FixedPosition = {
                left: 0,
                top: 0,
                width: 10,
                height: 10
            }

            const rect2: FixedPosition = {
                left: 5,
                top: 5,
                width: 10,
                height: 10
            }

            expect(isOverlapping(rect1, rect2)).toBeTruthy()
        })

        it("should show that rect2 and rect1 are overlapping", () => {
            const rect1: FixedPosition = {
                left: 0,
                top: 0,
                width: 10,
                height: 10
            }

            const rect2: FixedPosition = {
                left: 5,
                top: 5,
                width: 10,
                height: 10
            }

            expect(isOverlapping(rect2, rect1)).toBeTruthy()
        })

        it("should show that rect2 is next to rect1", () => {
            const rect1: FixedPosition = {
                left: 0,
                top: 0,
                width: 10,
                height: 10
            }

            const rect2: FixedPosition = {
                left: 20,
                top: 20,
                width: 10,
                height: 10
            }

            expect(isOverlapping(rect2, rect1)).toBeFalsy()
        })
    })
```